feat(benchmark/alloc): introduce minimal allocation benchmark

s1rk0 opened the pull request:
## Summary
This PR adds a microbenchmark for host memory allocation under benchmark/alloc/.  
It measures the cost of:
1) malloc/new (allocation),
2) first-use/touch (memset over the whole buffer), and
3) a trivial read (to prevent dead-store elimination),

so results reflect realistic “allocate + initialize + first access” latency rather than raw allocator timing.

## What’s included
- benchmark/alloc/CMakeLists.txt
- benchmark/alloc/src/alloc_benchmark.cpp
- benchmark/CMakeLists.txt: add_subdirectory(alloc/)

All new files carry SPDX headers.

## How to build & run
```bash
cmake -S . -B build -G Ninja \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
  -Dalpaka_BENCHMARKS=ON -Dalpaka_TESTING=OFF
cmake --build build -j
./build/benchmark/alloc/alloc
```

s1rk0 committed:
feat(benchmark/alloc): add memory-allocation micro-benchmark

* new file `benchmark/alloc/src/alloc_benchmark.cpp`
* new file `benchmark/alloc/CMakeLists.txt`
* updated root `benchmark/CMakeLists.txt`
  – `add_subdirectory(alloc/)`

s1rk0 committed:
perf(alloc): measure alloc + first-use (memset) instead of raw alloc/free

Switch benchmark to touch buffers after allocation:
- zero-initialize via queue memset
- read first byte to enforce observable use
This removes lazy-commit effects and makes results reflect
“allocate & start using” latency/bandwidth, not just allocator call overhead.
Discussed with Jiri on 2025-08-11.

s1rk0 committed:
fix(alloc): remove duplicated retval initialization

Drop accidental duplicate 'uint8_t retval' line in measureAlloc().

s1rk0 committed:
refactor(alloc): apply clang-format
- run clang-format to enforce style

s1rk0 committed:
chore(alloc): add author & funding headers

psychocoderHPC commented on benchmark/alloc/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+# Collect all .cpp files (change the glob if you move sources)
+file(GLOB_RECURSE _FILES_SOURCE "./src/*.cpp")
+
+alpaka_add_executable(${_TARGET_NAME} ${_FILES_SOURCE})
```
There was a change in alpaka with #230

please use here `add_executable(${_TARGET_NAME} ${_FILES_SOURCE})`

and 

```
target_link_libraries(${_TARGET_NAME} PRIVATE alpaka::alpaka)
alpaka_finalize(${_TARGET_NAME})
```

see ( note in one line there is a bug I found during copy past it into this post `PRIVATE PRIVATE` should be `PRIVATE`, I will fix it in alpaka soon) https://github.com/alpaka-group/alpaka3/blob/d310baba071f316c9927e6f5b49f5e14b5adb8ab/benchmark/babelstream/CMakeLists.txt#L24-L27C7

psychocoderHPC commented on benchmark/alloc/src/alloc_benchmark.cpp:
```diff
@@ -0,0 +1,360 @@
+    {
+        auto buf = fn(); // 1) allocate buffer
+        alpaka::onHost::memset(queue, buf, uint8_t{0}); // 2) touch buffer: zero-init
+        retval = static_cast<uint8_t>(buf[0]); // 3) read first byte to enforce use
```
`buf[0]` accessing the buffer is not allowed if you allocated for the GPU except you use `allocUnified()`
The line is not required  anyway because you touch the memory already with `onHost::memset()`

psychocoderHPC commented on benchmark/alloc/src/alloc_benchmark.cpp:
```diff
@@ -0,0 +1,360 @@
+}
+
+// ---------------- runtime container (2 "kernels": Host / Device) ----------
+struct AllocResults
```
could you move this and maybe other helpers into a header file to make it more simple to read the benchmark code only.

psychocoderHPC commented on benchmark/alloc/src/alloc_benchmark.cpp:
```diff
@@ -0,0 +1,360 @@
+};
+
+// ---------------- per‑backend test body -------------------------------------
+template<typename DeviceSpec, typename Exec>
```
please call it `T_DeviceSpec` and `T_Exec`. That's nothing you can know because this part of the documentation is not finished.

If the **template argument** is a type start always with `typename T_LargeCamelCase` and if it is an value than `uint32_t T_smallCamelCase`.

psychocoderHPC reviewed (changes requested):
Only small changes are required.

s1rk0 committed:
fix(alloc): sync backend helpers with upstream API

s1rk0 committed:
feat(alloc): backend-safe alloc timing with separate cold run; exclude cold from stats

s1rk0 committed:
refactor(alloc): extract benchmark helpers into shared header

s1rk0 commented:
Rebased on latest upstream/dev. Addressed review: switched to add_executable + alpaka_finalize, removed host deref of device buffers, moved helpers to alloc_helpers.hpp, renamed templates to T_*. Added cold/hot handling (first run reported separately and excluded from stats).

psychocoderHPC committed:
reformat the code with git pre-commit

psychocoderHPC commented:
@s1rk0 I reformated you PR follwing https://alpaka.readthedocs.io/en/latest/dev/style.html#pre-commit that the style checker is happy too. I will review you coode sasap
